// pino-logger-utils/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.logger = void 0;
require("./logger");
class Logger {
    constructor() {
        this.colors = {
            reset: '\x1b[0m',
            trace: '\x1b[90m',
            debug: '\x1b[36m',
            info: '\x1b[32m',
            warn: '\x1b[33m',
            error: '\x1b[31m',
            fatal: '\x1b[35m',
        };
        this.levelIcons = {
            trace: '🔍',
            debug: '🐛',
            info: 'ℹ️',
            warn: '⚠️',
            error: '❌',
            fatal: '💀',
        };
    }
    formatTimestamp() {
        const now = new Date();
        return now.toISOString();
    }
    formatMessage(level, message, options = {}) {
        const { timestamp = true, colorize = true } = options;
        const color = colorize ? this.colors[level] : '';
        const reset = colorize ? this.colors.reset : '';
        const icon = this.levelIcons[level];
        const time = timestamp ? `[${this.formatTimestamp()}]` : '';
        const levelStr = level.toUpperCase().padEnd(5);
        return `${color}${time} ${icon} ${levelStr}${reset} ${message}`;
    }
    trace(message, ...args) {
        console.log(this.formatMessage('trace', message), ...args);
    }
    debug(message, ...args) {
        console.log(this.formatMessage('debug', message), ...args);
    }
    info(message, ...args) {
        console.log(this.formatMessage('info', message), ...args);
    }
    warn(message, ...args) {
        console.warn(this.formatMessage('warn', message), ...args);
    }
    error(message, ...args) {
        console.error(this.formatMessage('error', message), ...args);
    }
    fatal(message, ...args) {
        console.error(this.formatMessage('fatal', message), ...args);
    }
    log(level, message, ...args) {
        switch (level) {
            case 'trace':
                this.trace(message, ...args);
                break;
            case 'debug':
                this.debug(message, ...args);
                break;
            case 'info':
                this.info(message, ...args);
                break;
            case 'warn':
                this.warn(message, ...args);
                break;
            case 'error':
                this.error(message, ...args);
                break;
            case 'fatal':
                this.fatal(message, ...args);
                break;
        }
    }
}
exports.logger = new Logger();
exports.default = exports.logger;
